sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,356 @@
+    def _set_angular_velocity(self):
+        pass
+
+    @abstractmethod
+    def _set_linear_velocity(self):
+        pass
+
+    def _generate_kdes(self):
+        kdes = []
+        t = dynamicsymbols._t
+        for i in range(len(self._coordinates)):
+            kdes.append(-self._coordinates[i].diff(t) + self._speeds[i])
+        return kdes
+
+    def _axis(self, body, ax):
+        if ax is None:
+            ax = body.frame.x
+            return ax
+        if not isinstance(ax, Vector):
+            raise TypeError("Axis must be of type Vector. Example-> A.x wehere 'A' is ReferenceFrame.")
+        if not ax.dt(body.frame) == 0:
+            raise ValueError('Axis cannot be time-varying.')
+        return ax
+
+    def _locate_joint_pos(self, body, joint_pos):
+        if joint_pos is None:
+            joint_pos = Vector(0)
+        if not isinstance(joint_pos, Vector):
+            raise ValueError('Joint Position must be supplied as Vector.')
+        if not joint_pos.dt(body.frame) == 0:
+            raise ValueError('Position Vector cannot be time-varying.')
+        return body.masscenter.locatenew(self._name + '_' + body.name + '_joint', joint_pos)
+
+    def _align_axis(self, parent, child):
+        # Returns the axis and angle between two axis(vectors).
+        angle = parent.angle_between(child)
+        axis = cross(child, parent).normalize()
+        return angle, axis
+
+
+class PinJoint(Joint):
+    """Pin (Revolute) Joint.
+
+    Explanation
+    ===========
+
+    It is defined such that the child body rotates with respect to the parent body
+    about the body fixed parent axis through the angle theta. The point of joint
+    (pin joint) is defined by two points in each body which coincides together.
+    In addition, the child's reference frame can be arbitrarily rotated a constant
+    amount with respect to the parent axis by passing an axis in child body which
+    must align with parent axis after the rotation.
+
+    Parameters
+    ==========
+
+    name : string
+        The Joint's name which makes it unique.
+    parent : Body
+        The parent body of joint.
+    child : Body
+        The child body of joint.
+    coordinates: dynamicsymbol, optional
+        Coordinates of joint.
+    speeds : dynamicsymbol, optional
+        Speeds of joint.
+    parent_joint_pos : Vector, optional
+        Defines the joint's point where the parent will be connected to child.
+        Default value is masscenter of Parent Body.
+    child_joint_pos : Vector, optional
+        Defines the joint's point where the child will be connected to parent.
+        Default value is masscenter of Child Body.
+    parent_axis : Vector, optional
+        Axis of parent frame which would be be aligned with child's
+        axis. Default is x axis in parent's frame.
+    child_axis : Vector, optional
+        Axis of child frame which would be aligned with parent's
+        axis. Default is x axis in child's frame.
```

Comment:
All of these descriptions can be improved by making them more precisely descriptive. The descriptions should match with a very nice generalized figure of the joint.